### ISSUE TITLE
Sequence length errors

### ISSUE THREAD

**Issue by kylebgorman** [CUNY-CL/yoyodyne#40]
This scary error sometimes greets you if you are using the transformer:

```
../aten/src/ATen/native/cuda/IndexKernel.cu:92: operator(): block: [398,0,0], thread: [51,0,0] Assertion `index >= -sizes[i] && index < sizes[i] && "index out of bounds"` failed.
```

I believe this is what happens when you have inputs that exceed `--max_sequence_length` (increasing it to a huge number makes it go away). What should we do about this? A few possibilities:

1. Cowardly refuse to pad a source sequence batch longer than `--max_sequence_length` by throwing an exception. This would have the interesting effect that it would apply not just to transformer anymore (which actually might be desirable---certainly it's clearer).
2. Document in the README that that scary errors means you need to increase `--max_sequence_length`.

I think I prefer (1). What do you think?

Relatedly, it occurs that the following might be clearer:

* `--max_sequence_length` > `--max_source_length`; help docs should also mention that this is for transformers only, unless we adopt (1) in which case it becomes uniform; we should also move this into base.py
* `--max_decode_length` > `--max_target_length`

**Comment by Adamits** [comment]
Does this raise an exception? Or is that message a warning? I believe that is just used to create the positional encoding tensor.

Anyway I also prefer 1). Though we could maybe remove it as an argument all together and just compute it dynamically before initializing the positional encoding. I would need to look into the implications of getting potentially huge positional encodings though.

**Comment by Adamits** [comment]
I **think** setting a `--max_sequence_length` is largely inspired by pretrained transformers that set such limits. This is probably for efficiency reasons, and during pretraining they just chunk documents to be within the threshold. However, I also imagine there is a point at which encoding performance also degrades if a sequence is too long. I don't really know the implications of this for character transduction.

I think if we keep this argument and raise an exception as you say, then that makes the choice of how to handle long sequences here explicit and up to the user.

**Comment by kylebgorman** [comment]
> Does this raise an exception? Or is that message a warning? I believe that is just used to create the positional encoding tensor.

It's a warning, but I think it should be raised to an exception, since it's a major issue.

> Anyway I also prefer 1). Though we could maybe remove it as an argument all together and just compute it dynamically before initializing the positional encoding. I would need to look into the implications of getting potentially huge positional encodings though.

That could be done later, maybe. I don't know if you can actually know how long the positional encoding needs to be at the time we're constructing the transformer model instance. 

> I think setting a --max_sequence_length is largely inspired by pretrained transformers that set such limits. This is probably for efficiency reasons, and during pretraining they just chunk documents to be within the threshold. However, I also imagine there is a point at which encoding performance also degrades if a sequence is too long. I don't really know the implications of this for character transduction.

**Comment by Adamits** [comment]
> * `--max_sequence_length` > `--max_source_length`; help docs should also mention that this is for transformers only, unless we adopt (1) in which case it becomes uniform; we should also move this into base.py
> * `--max_decode_length` > `--max_target_length`

These sound like good changes to me. I had thought of `--max_decode_length` as a cutoff for decoding without teacher forcing, in case the model starts doing that thing where it repeats the same character and never decodes and <EOS>. However, we may want to cutoff the length of target strings in the train data (or more likely raise an exception) instead. 

It seems that `transformer`/`transducer` do this slightly differently than the `LSTM` variants, which may be a bug.

**Comment by kylebgorman** [comment]
I don't really remember how `--max_decode_length` works so I'll defer to you there.

**Comment by kylebgorman** [comment]
@Adamits do you think we should document that this is LSTM (or "LSTM backend") only? Or should we enforce it on the other models too?

**Comment by Adamits** [comment]
Actually I misremembered slightly. For LSTMs we decode up to:

```
target.size(1) if target is not None else self.max_decode_length
```

This means that during training we will always decode up to the target length. I think every model should work the same wrt this hyperparameter. The LSTM behavior should be slightly more efficient (avoid useless decodings during training). Though then I guess the only increase in loss for not decoding EOS at the right timestep is from that single timestep -- not from further decoding. I am also unsure of how this works with the transformer which uses an MRT optimization AFAIK.

Furthermore I think I need to test what is happening now -- come to think of it I would expect the loss function to have sometimes errored for `Transformer` since it could end up with a longer prediction tensor than target tensor, and I did not think this got truncated by default.

**Comment by kylebgorman** [comment]
Okay, I'll assign this to you then.

**Comment by Adamits** [comment]
> Okay, I'll assign this to you then.

Sounds good!

**Comment by Adamits** [comment]
`max_target_length` behavior does, in fact make sense. The reason it is different for Transformer is because we have a separate `_decode_greedy` method, and `decode` relies on the pytorch `TransformerDecoder` behavior. The other models dynamically decide whether to decode greedily or with teacher forcing.

We should consider making this consistent, but it is really low priority imo.

The PR I opened just adds some checks on the source/target lengths in the collator.

**Comment by kylebgorman** [comment]
Closed by #53.
